File: MalDB.py
import sqlite3
from Bimap import Bimap
import pysparse

class DBWriter:
    """
    This class provides an interface for adding entries into the database.
    Changes made are not saved until saveChanges() is called. 
    """
    
    def __init__(self, dbName):
        self.conn = sqlite3.connect(dbName)
        
    def addUser(self, username, id):
        self.conn.cursor().execute("insert or ignore into userIds('id','name') \
                                    values(?,?)", (id, username))
        
    def addAnime(self, animename, id):
        self.conn.cursor().execute("insert or ignore into \
                     animeIds('id','name') values(?,?)", (id, animename))
        
    def addAnimeRating(self, userId, animeId, rating):
        self.conn.cursor().execute("insert or replace into ratings('userId',\
               'animeId','rating') values(?,?,?)", (userId, animeId, rating))
        
    def saveChanges(self):
        self.conn.commit()
            
class MemReader:
    """
    A memory reader for the sqlite3 database. It reads data from the database 
    and stores a copy in memory, which can then be quickly accessed without 
    accessing the hard drive. This does mean that changes made afterward to the
    database are not viewable from the memory reader.

    Also, anime and users are typically referred to their index in the 
    list of animes and users, rather than the UID of the site. This makes
    it easier to construct the ratings matrix.
    """

    def __init__(self, dbName):
        conn = sqlite3.connect(dbName)
                
        # Read the anime ids and names 
        c = conn.cursor()
        c.execute("select id, name from animeIds")

        animeIdStrings = [(row[0], row[1]) for row in c]
        animeIdIndices = [(pair[0], i) for i, pair in\
                           enumerate(animeIdStrings)]
        animeStringMap = Bimap(animeIdStrings, enforceBijection = False)
        animeIndexMap = Bimap(animeIdIndices, enforceBijection = False)
        c.close()

        # Read the user ids and names
        c = conn.cursor()
        c.execute("select id, name from userIds")

        userIdStrings = [(row[0], row[1]) for row in c]
        userIdIndices = [(pair[0], i) for i, pair in\
                           enumerate(userIdStrings)]
        userStringMap = Bimap(userIdStrings, enforceBijection = False)
        userIndexMap = Bimap(userIdIndices, enforceBijection = False)
        c.close()

        # Read all ratings triples and feed them into the matrix
        c = conn.cursor()
        c.execute("select userId, animeId, rating from ratings where rating \
                   <> 0")

        self.ratingsMatrix = pysparse.spmatrix.ll_mat(len(userIdStrings),
                                                      len(animeIdStrings))
        for row in c:
            userId = row[0]
            animeId = row[1]
            rating = row[2]
            
            userIndex  = userIndexMap.value(userId)
            animeIndex = animeIndexMap.value(animeId)
            self.ratingsMatrix[userIndex, animeIndex] = rating
            
        c.close()

        # Now create a bimap between indices and names directly, so that
        # the ids are no longer needed.
        userIndexStrings = [(index, userStringMap.value(\
                                    userIndexMap.key(index)))\
                            for index in range(len(userIdStrings))]

        animeIndexStrings = [(index, animeStringMap.value(\
                              animeIndexMap.key(index)))\
                            for index in range(len(animeIdStrings))]

        self.userMap = Bimap(userIndexStrings, enforceBijection = False)
        self.animeMap = Bimap(animeIndexStrings, enforceBijection = False)
        self.numUsers = len(userIdStrings)
        self.numAnime = len(animeIdStrings)


    def getUserIndex(self, username):
        return self.userMap.key(username)
        
    def getAnimeIndex(self, animename):
        return self.animeMap.key(animename)
        
    def getNumUsers(self):
        return self.numUsers
        
    def getNumAnime(self):
        return self.numAnime

    def getUserName(self, userIndex):
        return self.userMap.value(userIndex)
        
    def getAnimeName(self, animeIndex):
        return self.animeMap.value(animeIndex)

    def getAnimeRatingsForUser(self, userIndex):
        userMat = self.ratingsMatrix[userIndex, :]
        
        nonZeroRatings = [(col, rating) for ((row, col), rating) in\
                          userMat.items()]
 
        return dict(nonZeroRatings) 


    def getUserRatingsForAnime(self, animeIndex):
        animeMat = self.ratingsMatrix[:, animeIndex]
        
        nonZeroRatings = [(row, rating) for ((row, col), rating) in\
                          animeMat.items()]

        return dict(nonZeroRatings)
 
    def getRatingsMatrix(self):
        return self.ratingsMatrix


    

    
            
        
            
            
        

File: MalDB.cpp
#include "MalDB.h"
#include <sqlite3.h>
#include <stdlib.h>
#include <assert.h>
#include <cstdlib>

MalDBReader :: MalDBReader(string filename)
{
    sqlite3_open(filename.c_str(), &dbConn);

    int numRows;
    int numCols;
    char * errorString;
    char ** results;
    string query;

	map<unsigned int, unsigned int> userIdToIndex;
	map<unsigned int, unsigned int> animeIdToIndex;

	// Load up user ids
    query = string("select id, name from userIds");

    sqlite3_get_table(dbConn, query.c_str(), &results, &numRows,
                                &numCols, &errorString);

    for (int i = 1; i <= numRows; i++) 
    {
        unsigned int id = atoi(results[i * numCols]);
        string name = string(results[i * numCols + 1]);

        usernames.push_back(name);
		userIdToIndex.insert(pair<unsigned int, unsigned int>(id, i -1));
    }

	sqlite3_free_table(results);


	// Load up anime ids

    query = string("select id, name from animeIds");

    sqlite3_get_table(dbConn, query.c_str(), &results, &numRows,
                                &numCols, &errorString);

    for (int i = 1; i <= numRows; i++) 
    {
        unsigned int id = atoi(results[i * numCols]);
        string name = string(results[i * numCols + 1]);

        animenames.push_back(name);
		animeIdToIndex.insert(pair<unsigned int, unsigned int>(id, i -1));
    }

	sqlite3_free_table(results);

	// Load up ratings

	userToAnimeRated.resize(usernames.size());
	ratingsMatrix.resize(usernames.size());

	animeToUserRated.resize(animenames.size());
	ratingsMatrixTranspose.resize(animenames.size());
	
	query = string("select userid, animeid, rating from ratings where rating <> 0");

    sqlite3_get_table(dbConn, query.c_str(), &results, &numRows,
                                &numCols, &errorString);

    for (int i = 1; i <= numRows; i++) 
    {
        unsigned int user  = atoi(results[i * numCols]);
		unsigned int anime = atoi(results[i * numCols + 1]);
        double rating = strtod(results[i * numCols + 2], NULL);

		unsigned int userIndex  = userIdToIndex.find(user)  -> second;
		unsigned int animeIndex = animeIdToIndex.find(anime) -> second;

        userToAnimeRated[userIndex].push_back(animeIndex);
		ratingsMatrix[userIndex].push_back(rating);

        animeToUserRated[animeIndex].push_back(userIndex);
		ratingsMatrixTranspose[animeIndex].push_back(rating);

    }

	sqlite3_free_table(results);
	
}

MalDBReader :: ~MalDBReader()
{
	sqlite3_close(dbConn);

}


unsigned int MalDBReader :: getUserId(const string& name)
{
	int index = 0;
	for (; usernames[index] != name && index < usernames.size(); index++);

	assert(index <= usernames.size());
	
	return index;
}

unsigned int MalDBReader :: getAnimeId(const string& name)
{
	int index = 0;
	for (; animenames[index] != name && index < animenames.size(); index++);

	assert(index <= animenames.size());
	
	return index;
}

unsigned int MalDBReader :: getNumUsers()
{
	return usernames.size();
}

unsigned int MalDBReader :: getAnimeUsers()
{
	return animenames.size();
}

        
string& MalDBReader :: getUserName(unsigned int userId)
{
	return usernames[userId];
}
        
string& MalDBReader :: getAnimeName(unsigned int animeId)
{
    return animenames[animeId];
}

const vector<double>& MalDBReader :: getAnimeRatings(unsigned int userId)
{
	return ratingsMatrix[userId];
}

const vector<double>& MalDBReader :: getUserRatings(unsigned int animeId)
{
	return ratingsMatrixTranspose[animeId];
}

const vector<unsigned int>& MalDBReader :: getAnimeRated(unsigned int userId)
{
	return userToAnimeRated[userId];
}

const vector<unsigned int>& MalDBReader :: getUserRaters(unsigned int animeId)
{
	return animeToUserRated[animeId];
}

const vector< vector<double> >& MalDBReader :: getRatingsMatrix()
{
    return ratingsMatrix;
}

const vector< vector<double> >& MalDBReader :: getRatingsMatrixTranspose()
{
    return ratingsMatrixTranspose;
}

const vector<vector <unsigned int> >& MalDBReader :: getAnimeRatedList()
{
    return userToAnimeRated;
}

const vector<vector <unsigned int> >& MalDBReader :: getUserRatersList()
{
    return animeToUserRated;
}
